core/instantiate.ml

```diff
@@ -298,22 +298,32 @@ let datatype = instantiate_datatype
 
 module SEnv = Env.String
 
-let apply_type : Types.datatype -> Types.type_arg list -> Types.datatype =
-  fun pt tyargs ->
+let type_arguments_to_instantiation_maps :
+      bool -> Types.datatype -> Types.type_arg list -> (datatype * (datatype IntMap.t * row IntMap.t * field_spec IntMap.t)) =
```

Comment:
Consider introducing a type alias for `datatype IntMap.t * row IntMap.t * field_spec IntMap.t` to give it an informative name.